### DISCUSSION TITLE
Help to start the environment

### DISCUSSION THREAD

**Discussion by Mirlaaa** [plausible/analytics#3457] [Self-Hosted Support] [answered]
I've start to study the Elixir project as my first contact to technology.  Workin in self-hosting deployment... I'm facing problems when configuring my environment of development, wher run `make install`, a compilation error is raised:

== Compilation error in file lib/public_suffix.ex ==
** (MatchError) no match of right hand side value: ["// This Source Code Form is subject to the terms of the Mozilla Public\r\n// License, v. 2.0. If a copy of the MPL was not distributed with this\r\n// file, You can obtain one at https://mozilla.org/MPL/2.0/.\r\n\r\n// Please pull this list from, and only from https://publicsuffix.org/list/public_suffix_list.dat,\r\n// rather than any other VCS sites. Pulling from any other URL is not guaranteed to be supported.\r\n\r\n// Instructions on pulling and using this list can be found at https://publicsuffix.org/list/.\r\n\r\n// ===BEGIN ICANN DOMAINS===\r\n\r\n// ac : http://nic.ac/rules.htm\r\nac\r\ncom.ac\r\nedu.ac\r\ngov.ac\r\nnet.ac\r\nmil.ac\r\norg.ac\r\n\r\n// ad : https://en.wikipedia.org/wiki/.ad\r\nad\r\nnom.ad\r\n\r\n// ae : https://tdra.gov.ae/en/aeda/ae-policies\r\nae\r\nco.ae\r\nnet.ae\r\norg.ae\r\nsch.ae\r\nac.ae\r\ngov.ae\r\nmil.ae\r\n\r\n// aero : see https://www.information.aero/index.php?id=66\r\naero\r\naccident-investigation.aero\r\naccident-prevention.aero\r\naerobatic.aero\r\naeroclub.aero\r\naerodrome.aero\r\nagents.aero\r\naircraft.aero\r\nairline.aero\r\nairport.aero\r\nair-surveillance.aero\r\nairtraffic.aero\r\nair-traffic-control.aero\r\nambulance.aero\r\namusement.aero\r\nassociation.aero\r\nauthor.aero\r\nballooning.aero\r\nbroker.aero\r\ncaa.aero\r\ncargo.aero\r\ncatering.aero\r\ncertification.aero\r\nchampionship.aero\r\ncharter.aero\r\ncivilaviation.aero\r\nclub.aero\r\nconference.aero\r\nconsultant.aero\r\nconsulting.aero\r\ncontrol.aero\r\ncouncil.aero\r\ncrew.aero\r\ndesign.aero\r\ndgca.aero\r\neducator.aero\r\nemergency.aero\r\nengine.aero\r\nengineer.aero\r\nentertainment.aero\r\nequipment.aero\r\nexchange.aero\r\nexpress.aero\r\nfederation.aero\r\nflight.aero\r\nfuel.aero\r\ngliding.aero\r\ngovernment.aero\r\ngroundhandling.aero\r\ngroup.aero\r\nhanggliding.aero\r\nhomebuilt.aero\r\ninsurance.aero\r\njournal.aero\r\njournalist.aero\r\nleasing.aero\r\nlogistics.aero\r\nmagazine.aero\r\nmaintenance.aero\r\nmedia.aero\r\nmicrolight.aero\r\nmodelling.aero\r\nnavigation.aero\r\nparachuting.aero\r\nparagliding.aero\r\npassenger-association.aero\r\npilot.aero\r\npress.aero\r\nproduction.aero\r\nrecreation.aero\r\nrepbody.aero\r\nres.aero\r\nresearch.aero\r\nrotorcraft.aero\r\nsafety.aero\r\nscientist.aero\r\nservices.aero\r\nshow.aero\r\nskydiving.aero\r\nsoftware.aero\r\nstudent.aero\r\ntrader.aero\r\ntrading.aero\r\ntrainer.aero\r\nunion.aero\r\nworkinggroup.aero\r\nworks.aero\r\n\r\n// af : http://www.nic.af/help.jsp\r\naf\r\ngov.af\r\ncom.af\r\norg.af\r\nnet.af\r\nedu.af\r\n\r\n// ag : http://www.nic.ag/prices.htm\r\nag\r\ncom.ag\r\norg.ag\r\nnet.ag\r\nco.ag\r\nnom.ag\r\n\r\n// ai : http://nic.com.ai/\r\nai\r\noff.ai\r\ncom.ai\r\nnet.ai\r\norg.ai\r\n\r\n// al : http://www.ert.gov.al/ert_alb/faq_det.html?Id=31\r\nal\r\ncom.al\r\nedu.al\r\ngov.al\r\nmil.al\r\nnet.al\r\norg.al\r\n\r\n// am : https://www.amnic.net/policy/en/Policy_EN.pdf\r\nam\r\nco.am\r\ncom.am\r\ncommune.am\r\nnet.am\r\norg.am\r\n\r\n// ao : https://en.wikipedia.org/wiki/.ao\r\n// http://www.dns.ao/REGISTR.DOC\r\nao\r\ned.ao\r\ngv.ao\r\nog.ao\r\nco.ao\r\npb.ao\r\nit.ao\r\n\r\n// aq : https://en.wikipedia.org/wiki/.aq\r\naq\r\n\r\n// ar : https://nic.ar/es/nic-argentina/normativa\r\nar\r\nbet.ar\r\ncom.ar\r\ncoop.ar\r\nedu.ar\r\ngob.ar\r\ngov.ar\r\nint.ar\r\nmil.ar\r\nmusica.ar\r\nmutual.ar\r\nnet.ar\r\norg.ar\r\nsenasa.ar\r\ntur.ar\r\n\r\n// arpa : https://en.wikipedia.org/wiki/.arpa\r\n// Confirmed by registry <iana-questions@icann.org> 2008-06-18\r\narpa\r\ne164.arpa\r\nin-addr.arpa\r\nip6.arpa\r\niris.arpa\r\nuri.arpa\r\nurn.arpa\r\n\r\n// as : https://en.wikipedia.org/wiki/.as\r\nas\r\ngov.as\r\n\r\n// asia : https://en.wikipedia.org/wiki/.asia\r\nasia\r\n\r\n// at : https://en.wikipedia.org/wiki/.at\r\n// Confirmed by registry <it@nic.at> 2008-06-17\r\nat\r\nac.at\r\nco.at\r\ngv.at\r\nor.at\r\nsth.ac.at\r\n\r\n// au : https://en.wikipedia.org/wiki/.au\r\n// http://www.auda.org.au/\r\nau\r\n// 2LDs\r\ncom.au\r\nnet.au\r\norg.au\r\nedu.au\r\ngov.au\r\nasn.au\r\nid.au\r\n// Historic 2LDs (closed to new registration, but sites still exist)\r\ninfo.au\r\nconf.au\r\noz.au\r\n// CGDNs - http://www.cgdn.org.au/\r\nact.au\r\nnsw.au\r\nnt.au\r\nqld.au\r\nsa.au\r\ntas.au\r\nvic.au\r\nwa.au\r\n// 3LDs\r\nact.edu.au\r\ncatholic.edu.au\r\n// eq.edu.au - Removed at the request of the Queensland Department of Education\r\nnsw.edu.au\r\nnt.edu.au\r\nqld.edu.au\r\nsa.edu.au\r\ntas.edu.au\r\nvic.edu.au\r\nwa.edu.au\r\n// act.gov.au  Bug 984824 - Removed at request of Greg Tankard\r\n// nsw.gov.au  Bug 547985 - Removed at request of <Shae.Do" <> ...]
    lib/public_suffix/rules_parser.ex:14: PublicSuffix.RulesParser.parse_rules/1
    lib/public_suffix.ex:195: (module)
could not compile dependency :public_suffix, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile public_suffix", update it with "mix deps.update public_suffix" or clean it with "mix deps.clean public_suffix"
make: *** [Makefile:8: install] Error 1

Someone to help me, please!?

**Comment by ruslandoga** [comment]
👋 @Mirlaaa 

The error happens here: https://github.com/axelson/publicsuffix-elixir/blob/fa40c243d4b5d8598b90cff268bc4e33f3bb63f1/lib/public_suffix/rules_parser.ex#L14-L16

I wonder what Elixir / Erlang / publicsuffix versions you are using? Here're mine and the compilation works for me:

```console
$ elixir --version
elixir --version
Erlang/OTP 25 [erts-13.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]
Elixir 1.14.3 (compiled with Erlang/OTP 25)
$ cat mix.lock | grep public_suffix
  "public_suffix": {:git, "https://github.com/axelson/publicsuffix-elixir", "89372422ab8b433de508519ef474e39699fd11ca", []},
```

I think it should be OK if you have newer Elixir / Erlang.

**Reply by ruslandoga** [reply]
Newer version of `public_suffix` seems to be working too:

```console
$ mix deps.update public_suffix
* Updating public_suffix (https://github.com/axelson/publicsuffix-elixir)
origin/HEAD set to main
Resolving Hex dependencies...
Resolution completed in 0.738s
$ mix deps.compile public_suffix
==> public_suffix
Compiling 3 files (.ex)
Generated public_suffix app
$ mix compile
Compiling 17 files (.ex)
Generated plausible app
$ cat mix.lock | grep public_suffix
  "public_suffix": {:git, "https://github.com/axelson/publicsuffix-elixir", "fa40c243d4b5d8598b90cff268bc4e33f3bb63f1", []},
```

**Reply by Mirlaaa** [reply]
After cleaning, updating and compiling the dependency again, the error persists.

**Reply by ruslandoga** [reply]
Hm. I guess the problem is with the "rule file". https://github.com/axelson/publicsuffix-elixir/blob/fa40c243d4b5d8598b90cff268bc4e33f3bb63f1/lib/public_suffix/rules_parser.ex#L14-L16 shows that the files is expected to have `===END ICANN DOMAINS===\n// ===BEGIN PRIVATE DOMAINS===` in it, but the error you are seeing only gets a single value in response, meaning there is no such string found.

**Reply by ruslandoga** [reply]
I see `===BEGIN ICANN DOMAINS===` in your error message, though.

**Reply by ruslandoga** [reply]
I guess it's about `\r\n` vs `\n`. `publicsuffix` expects there to be just newlines `\n`, and your error shows that in your case there are also carriage returns `\r`. What operating system are you using? This seems a bit like a Windows thing.

On my machine it's like this:

```elixir
iex> Mix.install [:req]
iex> Req.get!("https://publicsuffix.org/list/public_suffix_list.dat").body
"// This Source Code Form is subject to the terms of the Mozilla Public\n// License, v. 2.0. If a copy of the MPL was not distributed with this\n// file, You can obtain one at https://mozilla.org/MPL/2.0/.\n\n// Please pull this list from, and only from https://publicsuffix.org/list/public_suffix_list.dat,\n// rather than any other VCS sites. Pulling from any other URL is not guaranteed to be supported.\n\n// Instructions on pulling and using this list can be found at https://publicsuffix.org/list/.\n\n// ===BEGIN ICANN DOMAINS===\n\n// ac : http://nic.ac/rules.htm\nac\ncom.ac\nedu.ac\ngov.ac\nnet.ac\nmil.ac\norg.ac\n\n// ad : https://en.wikipedia.org/wiki/.ad\nad\nnom.ad\n\n// ae : https://tdra.gov.ae/en/aeda/ae-policies\nae\nco.ae\nnet.ae\norg.ae\nsch.ae\nac.ae\ngov.ae\nmil.ae\n\n// aero : see https://www.information.aero/index.php?id=66\naero\naccident-investigation.aero\naccident-prevention.aero\naerobatic.aero\naeroclub.aero\naerodrome.aero\nagents.aero\naircraft.aero\nairline.aero\nairport.aero\nair-surveillance.aero\nairtraffic.aero\nair-traffic-control.aero\nambulance.aero\namusement.aero\nassociation.aero\nauthor.aero\nballooning.aero\nbroker.aero\ncaa.aero\ncargo.aero\ncatering.aero\ncertification.aero\nchampionship.aero\ncharter.aero\ncivilaviation.aero\nclub.aero\nconference.aero\nconsultant.aero\nconsulting.aero\ncontrol.aero\ncouncil.aero\ncrew.aero\ndesign.aero\ndgca.aero\neducator.aero\nemergency.aero\nengine.aero\nengineer.aero\nentertainment.aero\nequipment.aero\nexchange.aero\nexpress.aero\nfederation.aero\nflight.aero\nfuel.aero\ngliding.aero\ngovernment.aero\ngroundhandling.aero\ngroup.aero\nhanggliding.aero\nhomebuilt.aero\ninsurance.aero\njournal.aero\njournalist.aero\nleasing.aero\nlogistics.aero\nmagazine.aero\nmaintenance.aero\nmedia.aero\nmicrolight.aero\nmodelling.aero\nnavigation.aero\nparachuting.aero\nparagliding.aero\npassenger-association.aero\npilot.aero\npress.aero\nproduction.aero\nrecreation.aero\nrepbody.aero\nres.aero\nresearch.aero\nrotorcraft.aero\nsafety.aero\nscientist.aero\nservices.aero\nshow.aero\nskydiving.aero\nsoftware.aero\nstudent.aero\ntrader.aero\ntrading.aero\ntrainer.aero\nunion.aero\nworkinggroup.aero\nworks.aero\n\n// af : http://www.nic.af/help.jsp\naf\ngov.af\ncom.af\norg.af\nnet.af\nedu.af\n\n// ag : http://www.nic.ag/prices.htm\nag\ncom.ag\norg.ag\nnet.ag\nco.ag\nnom.ag\n\n// ai : http://nic.com.ai/\nai\noff.ai\ncom.ai\nnet.ai\norg.ai\n\n// al : http://www.ert.gov.al/ert_alb/faq_det.html?Id=31\nal\ncom.al\nedu.al\ngov.al\nmil.al\nnet.al\norg.al\n\n// am : https://www.amnic.net/policy/en/Policy_EN.pdf\nam\nco.am\ncom.am\ncommune.am\nnet.am\norg.am\n\n// ao : https://en.wikipedia.org/wiki/.ao\n// http://www.dns.ao/REGISTR.DOC\nao\ned.ao\ngv.ao\nog.ao\nco.ao\npb.ao\nit.ao\n\n// aq : https://en.wikipedia.org/wiki/.aq\naq\n\n// ar : https://nic.ar/es/nic-argentina/normativa\nar\nbet.ar\ncom.ar\ncoop.ar\nedu.ar\ngob.ar\ngov.ar\nint.ar\nmil.ar\nmusica.ar\nmutual.ar\nnet.ar\norg.ar\nsenasa.ar\ntur.ar\n\n// arpa : https://en.wikipedia.org/wiki/.arpa\n// Confirmed by registry <iana-questions@icann.org> 2008-06-18\narpa\ne164.arpa\nin-addr.arpa\nip6.arpa\niris.arpa\nuri.arpa\nurn.arpa\n\n// as : https://en.wikipedia.org/wiki/.as\nas\ngov.as\n\n// asia : https://en.wikipedia.org/wiki/.asia\nasia\n\n// at : https://en.wikipedia.org/wiki/.at\n// Confirmed by registry <it@nic.at> 2008-06-17\nat\nac.at\nco.at\ngv.at\nor.at\nsth.ac.at\n\n// au : https://en.wikipedia.org/wiki/.au\n// http://www.auda.org.au/\nau\n// 2LDs\ncom.au\nnet.au\norg.au\nedu.au\ngov.au\nasn.au\nid.au\n// Historic 2LDs (closed to new registration, but sites still exist)\ninfo.au\nconf.au\noz.au\n// CGDNs - http://www.cgdn.org.au/\nact.au\nnsw.au\nnt.au\nqld.au\nsa.au\ntas.au\nvic.au\nwa.au\n// 3LDs\nact.edu.au\ncatholic.edu.au\n// eq.edu.au - Removed at the request of the Queensland Department of Education\nnsw.edu.au\nnt.edu.au\nqld.edu.au\nsa.edu.au\ntas.edu.au\nvic.edu.au\nwa.edu.au\n// act.gov.au  Bug 984824 - Removed at request of Greg Tankard\n// nsw.gov.au  Bug 547985 - Removed at request of <Shae.Donelan@services.nsw.gov.au>\n// nt.gov.au  Bug 940478 - Removed at request of Greg Connors <Greg.Connors@nt.gov.au>\nqld.gov.au\nsa.gov.au\ntas.gov.au\nvic.gov.au\nwa.gov.au\n// 4LDs\n// education.tas.edu.au - Removed at the request of the Department of Education " <> ...
```

**Reply by Mirlaaa** [reply]
Currently, using Windows 11.

**Reply by ruslandoga** [reply, marked as answer]
I don't think anyone on the core team is using Windows, so there might be small problems like this one here and there. I'd suggest you consider using Linux Windows Subsystem to build and develop Plausible :)

**Reply by Mirlaaa** [reply]
Ooh, nice! 
Thanks for helping me, I'm gonna start to working on this. 🤝

**Comment by Mirlaaa** [comment]
> 👋 @Mirlaaa
> 
> The error happens here: https://github.com/axelson/publicsuffix-elixir/blob/fa40c243d4b5d8598b90cff268bc4e33f3bb63f1/lib/public_suffix/rules_parser.ex#L14-L16
> 
> I wonder what Elixir / Erlang / publicsuffix versions you are using? Here're mine and the compilation works for me:
> 
> ```
> $ elixir --version
> elixir --version
> Erlang/OTP 25 [erts-13.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]
> Elixir 1.14.3 (compiled with Erlang/OTP 25)
> $ cat mix.lock | grep public_suffix
>   "public_suffix": {:git, "https://github.com/axelson/publicsuffix-elixir", "89372422ab8b433de508519ef474e39699fd11ca", []},
> ```
> 
> I think it should be OK if you have newer Elixir / Erlang.


These are the versions:
```
elixir --version && cat .\mix.lock | grep public_suffix
Erlang/OTP 25 [erts-13.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit:ns]

Elixir 1.14.3 (compiled with Erlang/OTP 25)
  "public_suffix": {:git, "https://github.com/axelson/publicsuffix-elixir", "fa40c243d4b5d8598b90cff268bc4e33f3bb63f1", []},
```
